File: 0078-subsets/0078-subsets.py
class Solution:
    def _dfs(self, nums, index, path, res) -> None:
        # Add the current path to the result
        res.append(path.copy())
        
        # Explore all possible subsets starting from the current index
        for i in range(index, len(nums)):
            path.append(nums[i])
            self._dfs(nums, i + 1, path, res)
            path.pop()  # Backtracking

    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        self._dfs(nums, 0, [], res)
        return res

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        dfs(nums, 0, path, res);
        return res;
    }
private:
    void dfs(vector<int>& nums, int idx, vector<int>& path, vector<vector<int>>& res) {
        if (idx >= nums.size()) {
            res.push_back(path);            
            return;
        }
        // decision to include nums[i]
        path.push_back(nums[idx]);
        dfs(nums, idx+1, path, res);
        
        // decision not to include nums[i]
        path.pop_back();
        dfs(nums, idx+1, path, res);
    }
};